internal/charm/charmdir_test.go

```diff
@@ -598,8 +599,9 @@ func (s *CharmSuite) TestMaybeGenerateVersionStringError(c *gc.C) {
 	dir, err := charm.ReadCharmDir(charmDir)
 	c.Assert(err, jc.ErrorIsNil)
 
-	version, vcsType, err := dir.MaybeGenerateVersionString()
-	msg := fmt.Sprintf("%q version string generation failed : exit status 128\nThis means that the charm version won't show in juju status. Charm path %q", "git", dir.Path)
+	versionReader := charm.DefaultVersionReader(loggertesting.WrapCheckLog(c))
+	version, vcsType, err := dir.MaybeGenerateVersionString(context.Background(), versionReader)
+	msg := fmt.Sprintf("failed reading version from %q: git version string generation failed : exit status 128\nThis means that the charm version won't show in juju status.", dir.Path)
```

Comment:
We're back to comparing error strings?

Reply:
I think in this refactor, yes. In reality, if we remove deploy from directory, all of this code can go.

Reply:
Actually, I have made a typed error.